File: 623-add-one-row-to-tree/623-add-one-row-to-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    
    def addNode(self, val, leftChild, rightChild):
        newNode = TreeNode(val, leftChild, rightChild);
        return newNode
        
    def solve(self, root, val, depth):
        if root == None: return None
        if depth == 1:
            root.left = self.addNode(val, root.left, None)
            root.right = self.addNode(val, None, root.right)
        
        self.solve(root.right, val, depth-1)
        self.solve(root.left, val, depth-1)
    
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        if depth == 1: return self.addNode(val, root, None)
        self.solve(root, val, depth-1)
        return root

File: 623-add-one-row-to-tree/623-add-one-row-to-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    
    TreeNode* addNode(int val, TreeNode* leftChild, TreeNode* rightChild){
        TreeNode* newNode = new TreeNode(val);
        newNode->left = leftChild;
        newNode->right = rightChild;
        return newNode;
    }
    
    void solve(TreeNode* root, int val, int depth){
        
        if(root == NULL) return;
        if(depth == 1){
            root->left = addNode(val, root->left, NULL);
            root->right = addNode(val, NULL, root->right);
            return;
        }
        solve(root->left, val, depth-1);
        solve(root->right, val, depth-1);
        
    }
    
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
        if(depth == 1) return addNode(val, root, NULL);
        solve(root, val, depth-1);
        return root;
        
    }
};